src/Collections/Sequences/Seq.dfy

```diff
@@ -0,0 +1,720 @@
+    }
+  }
+
+  /* finds the index of a certain element in the sequence */
+  function method {:opaque} IndexOf<T(==)>(s: seq<T>, v: T): (i: nat)
+    requires v in s
+    ensures i < |s| && s[i] == v
+  {
+    if s[|s|-1] == v then |s| - 1 else IndexOf(s[..|s|-1], v)
+  }
+
+  /* finds the index of a certain element in the sequence if found */
+  function method {:opaque} IndexOfOption<T(==)>(s: seq<T>, v: T): (o: Option<nat>)
+    ensures if o.Some? then o.value < |s| && s[o.value] == v else v !in s
+  {
+    if |s| == 0 then None()
+    else if s[|s|-1] == v then Some(|s| - 1) else IndexOfOption(s[..|s|-1], v)
+  }
+
+  /* slices out a specific position's value from the sequence */
+  function method {:opaque} Remove<T>(s: seq<T>, pos: nat): (s': seq<T>)
+    requires pos < |s|
+    ensures |s'| == |s| - 1
+    ensures forall i {:trigger s'[i], s[i]} | 0 <= i < pos :: s'[i] == s[i]
+    ensures forall i {:trigger s'[i]} | pos <= i < |s| - 1 :: s'[i] == s[i+1]
+  {
+    s[..pos] + s[pos+1..]
+  }
+
+  /* slices out a specific value from the sequence */
+  function method {:opaque} RemoveValue<T(==)>(s: seq<T>, v: T): (s': seq<T>)
+    ensures v !in s ==> s == s'
+    ensures v in s ==> |multiset(s')| == |multiset(s)| - 1
+    ensures v in s ==> multiset(s')[v] == multiset(s)[v] - 1
+    ensures HasNoDuplicates(s) ==> HasNoDuplicates(s') && ToSet(s') == ToSet(s) - {v}
+  {
+    reveal HasNoDuplicates();
+    reveal ToSet();
+    if v !in s then s
+    else
+      var i := IndexOf(s, v);
+      assert s == s[..i] + [v] + s[i+1..];
+      s[..i] + s[i+1..]
+  }
+
+  /* slices out a specific value from the sequence if found */
+  function method {:opaque} RemoveValueOption<T(==)>(s: seq<T>, o: Option<T>): (s': seq<T>)
+    ensures o == None || (|s| == 0 && |s'| == 0) ||
+      (|s| > 0 && var v := o.UnwrapOr(s[0]);
+        v !in s ==> s == s'
+        && v in s ==> |multiset(s')| == |multiset(s)| - 1
+        && v in s ==> multiset(s')[v] == multiset(s)[v] - 1
+        && HasNoDuplicates(s) ==> HasNoDuplicates(s') && ToSet(s') == ToSet(s) - {v})
+  {
+    reveal HasNoDuplicates();
+    reveal ToSet();
+    match o
+    case None => s
+    case Some(v) =>
+      if v !in s then s else
+      var i := IndexOf(s, v);
+      assert s == s[..i] + [v] + s[i+1..];
+      s[..i] + s[i+1..]
+  }
+
+  /* inserts a certain value into a specified index of the sequence */
+  function method {:opaque} Insert<T>(s: seq<T>, a: T, pos: nat): seq<T>
+    requires pos <= |s|;
+    ensures |Insert(s, a, pos)| == |s| + 1;
+    ensures forall i {:trigger Insert(s, a, pos)[i], s[i]}:: 0 <= i < pos ==> Insert(s, a, pos)[i] == s[i];
+    ensures forall i {:trigger s[i]} :: pos <= i < |s| ==> Insert(s, a, pos)[i+1] == s[i];
+    ensures Insert(s, a, pos)[pos] == a;
+    ensures multiset(Insert(s, a, pos)) == multiset(s) + multiset{a}
+  {
+    assert s == s[..pos] + s[pos..];
+    s[..pos] + [a] + s[pos..]
+  }
+
+  function method {:opaque} Reverse<T>(s: seq<T>): seq<T>
```

Comment:
I suppose this is non-blocking, but seems like we should have `ensures forall i | 0 <= i < |s| :: Reverse(s)[i] == s[|s| - i - 1]`